### ISSUE TITLE
Pagination support

### ISSUE THREAD

**Issue by JaeAeich** [ga4gh/task-execution-schemas#193]
The current implementation of the `API` responses lacks adequate support for comprehensive pagination. The API solely offers a `next_page_token` as a cursor to navigate to the next page. However, to access the previous page, it necessitates `client-side` caching of the cursors. Additionally, essential pagination elements such as the total number of tasks and pages, are crucial for effective pagination.

**Comment by uniqueg** [comment]
Strongly agree with this. The current pagination is really only useful for limiting the number of tasks returned by `GET /tasks`. For effective pagination in a GUI client it's not very helpful.

Perhaps we could look at the GA4GH pagination style guide that is [currently being drafted up](https://docs.google.com/document/d/12Y0MgIRqaBt2CqzEscArVEkCS_8WHFRdQef9IFmEGVc/edit) by TASC in repsonse to this issue: https://github.com/ga4gh/TASC/issues/29). It lists a couple of solutions and discusses pros and cons.

**Comment by JaeAeich** [comment]
It'd be nice to see pagination where user can:
- Go forward and backward 
- Jump to pages

Maybe setting up an extra cursor `page_token_prev` and `page_num` could be a good place to start. Also the API should allow users to fetch a specific page in that manner, so that user can just lets say jump to page 4 skipping all the middle pages. Apart from that `total_results` and `total_page_number` would also be essential, which would allow the user to directly jump to the last page.

**Comment by uniqueg** [comment]
@andrewyatz: Do you think these use cases are covered in the TASC pagination guide?

**Comment by andrewyatz** [comment]
Pagination guide would suggest not using tokens for exactly these reasons and instead paging using counts/offsets. Tokens have been written up to prefer streaming & not that you cannot replay paginated streams (like in Kafka) it is implementation specific.

I'd like to know if this is a "good" solution to the above

**Comment by uniqueg** [comment]
Thanks @andrewyatz.

@JaeAeich: Would either of the two solutions described in the [GA4GH pagination guide draft](https://docs.google.com/document/d/1pu2icPEll-vueFcjCuUnjuAwEvtahOf_dyyaP5qqbCs/edit) address your use case? If not, please leave feedback :pray:

**Comment by JaeAeich** [comment]
> When datasets may change between requests, clients SHOULD be tolerant of skipped or redundant entries between pages and servers SHOULD consider pagination stability requirements when choosing to offer an page-based pagination scheme.

How would client be tolerant to the data its unaware of, ie, lets say I fetched page n, and while fetching page n + 1, the data set changes and an entry gets added to page n, this means I'll have a redundant entry which my client has already seen in page n.
We can have a workaround by implementing some kind of cache invalidation, but I think this should be the job of the server.

### Page based
> total (optional) - The total number of results available in the result set

Client typically show the range of "pages" a specific user will have, ie if I request 5 items per page, I will expect 100/5 = 20  (say total 100) "pages", ie when on page n, pagination will show something like `1,2 ... ->n<- .. 19, 20`. To get this I think in this approach client would want to have total items. This way client can calculate offset and properly paginate.

> Servers MAY return less than the page_size, even if it is not the last page

I understand this might not be a problem for GUI, as result `page_size` would never exceed 100, but just outta curiosity... why not a wait till a threshold and error if its too resource intensive?

### Token based
Response doesn't have `prev_page_token`, client would need or will have to cache these token, which shouldn't be its job.

---

### Use case visualization

![image](https://github.com/ga4gh/task-execution-schemas/assets/100477031/bddf7e5d-79f8-43e7-a5dd-1e7545e33636)

PS: I think page_based would be preferred but would require breaking changes in [cloud-component](https://github.com/elixir-cloud-aai/cloud-components/).

**Comment by andrewyatz** [comment]
Thanks for the insights here. It might be better to move some of these to the TASC repo but from the first points here

**Data tolerance**
This is a key issue in any pagination scheme. When researching this I found many technologies attempt to use temporal pagination or paginate through a cached result server side to avoid the issue all together. I think the impact of this depends on how much you can or cannot tolerate repeats in data responses and that does not feel like the pagination's guide to say this

**Pages available**
I didn't see anything in the examples we reviewed which did provide the page count and as you said it's typically calculated by the client. So we didn't include it

**`prev_page_token`**
The availability of that token would depend on the state the server knows. It might have zero idea about what the previous page was, which is why it was omitted from the recommended returns. If the token was predictable then this is a different matter but we pushed for obfuscated tokens (which of course could be based on known data). 

Do you think this is an issue?

**Comment by uniqueg** [comment]
I agree that the first point should be left to implementers and that the second point is unusual and can easily be calculated by the client from the total number of records and the page size.

I also agree that a token for the previous page may or may not be useful, sensible, possible for a given recommendation and so should not be recommneded as required. It could, however, be mentioned in a note or as an optional extension (it would be cool that all the implementations that make use of it use it in the same manner).

**Comment by andrewyatz** [comment]
This is great thank you. I've added these suggestions. Please see what you think especially as your comments are with respect to promoting consistency. Really appreciate this

**Comment by JaeAeich** [comment]
Hi, I'll just add this here after some more research and experience that I've gone through about pagination and consistency.

tl;dr: Implement different strategy for different endpoint for based on expected records at scale. Data should being sent inside a data wapper for single or collection of record helps extensions, consistency is case is great.


### **Pagination Strategies**

The choice in API design for collections is between **offset-based** and **cursor-based** pagination. This decision is a direct tradeoff between user convenience and system scalability.

  * **Offset-Based Pagination:** This is the familiar "page 1 of 10" model. It's user-friendly because clients can request a specific page number, and the API can return the total count of items. This works well for smaller, stable datasets.

    **Example Request:**
    `GET /v1/documents?page=3&page_size=25`

    **Example Response:**

    ```json
    {
      "data": [
        /* 25 document objects */
      ],
      "meta": {
        "total": 125,
        "page_size": 25,
        "page_number": 3
      }
    }
    ```

    **Problem:** The `total` count requires an expensive `COUNT(*)` database query. On a table with millions of rows, this can be a performance killer. Additionally, skipping records with `OFFSET` becomes progressively slower.

  * **Cursor-Based Pagination:** This approach is built for scale. Instead of a page number, a client receives an opaque token or cursor. To get the next set of results, they simply pass this cursor back to the API. The server then efficiently fetches records "after" the one the cursor points to.

    **Example Request:**
    `GET /v1/tasks?page_size=50&next_page_token=v1_1678886400000_12345`

    **Example Response:**

    ```json
    {
      "data": [
        /* 50 task objects */
      ],
      "meta": {
        "page_size": 50,
        "next_page_token": "v1_1678886400000_12345",
        "prev_page_token": "v1_1678886400000_67890"
      }
    }
    ```

    **Advantage:** This completely avoids the `COUNT(*)` query and the performance hit of deep offsets. The API doesn't know the total number of records, so it can't tell the client what "page 100" is, but it can efficiently serve the next batch of results, regardless of how many records exist. This is the **correct choice for endpoints with large, unbounded datasets**, like a feed of millions of tasks.

### **CoNsisTeNcy Wishlist**

  * **Standardized Response Envelope:** Every response from the API should follow a predictable structure. This allows client libraries and front-end code to handle data and metadata in a consistent manner, reducing boilerplate and potential errors.

    **Weird:**

    ```json
    { "tasks": [ /* ... */ ] }
    { "documents": [ /* ... */ ] }
    ```

    This requires a unique handler for each endpoint.

    **Correct:**

    ```json
    { "data": [ /* ... */ ] }
    ```

    This allows a single, reusable function to parse the main data payload regardless of the endpoint.

  * **Consolidated Metadata:** Put all non-data information in one place. A `meta` object is a great pattern for this, consolidating pagination tokens, counts, or other relevant information. This is much cleaner than, for example, putting some pagination info in the headers and some in the body.

  * **Naming Convention:**

      * **URLs:** Use **kebab-case** for URL paths (`/tool-classes`, `/service-info`).
      * **JSON Fields:** Consistentency with either **snake\_case** (`first_name`) or **camelCase** (`firstName`). While camelCase is common in JavaScript, snake\_case is widely used in other large ecosystems (Rust, Zig, Python, Stripe, AWS) and is often considered more readable.

**Comment by uniqueg** [comment]
Thanks @JaeAeich. Great points, although, to me, I think they're probaby something to bring to the [TASC](https://www.ga4gh.org/what-we-do/technical-alignment-subcommittee-tasc/) to discuss in order to get agreement for all standards, not just on TES

TASC also drafted the [pagination guide](https://github.com/ga4gh/TASC/blob/main/recommendations/API%20pagination%20guide.md). With regard to the pagination strategies you describe, is there anything that is _wrong_ in that guide, in your opinion?

Also, what do you think, @andrewyatz?

I also foresee some problems with your suggestions on standardized response envelopes and metadata containers, mostly with backward compatibility. The ecosystems around most of the GA4GH (Open)API standards are relatively small, so further fragmentation through additional releases, especially major version changes, is a serious problem. Probably more so than consistency.

Finally, on naming, I think we should generally follow the examples on the official OpenAPI documentation, which probably (didn't check) follows JSON Schema conventions (given that it's based on JSON Schema).
